File: src/analyze.ts
import { execFile } from "node:child_process";
import path from "node:path";
import { promisify } from "node:util";
import { discoverServices, type DiscoveryAttempt } from "./discover";
import { isDir } from "./discover/fs";
import { BlastGraph } from "./graph/graph";
import type { DiscoveryStrategy, Edge, Service } from "./graph/types";

const run = promisify(execFile);

export interface Analysis {
  /** `owner/repo` from the git remote, or the directory name. */
  repository: string;
  /** Absolute path that was analysed. */
  root: string;
  discovery: {
    strategy: DiscoveryStrategy | null;
    attempted: DiscoveryAttempt[];
  };
  graph: BlastGraph;
  /** Stage 3 fills this in. Until then the report says so plainly. */
  runtime: { connected: false };
}

export interface AnalysisJSON {
  repository: string;
  root: string;
  discovery: Analysis["discovery"];
  services: Service[];
  edges: Edge[];
  runtime: Analysis["runtime"];
}

/** Runs every built stage over a repository and returns the graph. */
export async function analyze(root: string): Promise<Analysis> {
  const absRoot = path.resolve(root);
  if (!(await isDir(absRoot))) throw new Error(`not a directory: ${root}`);

  const [repository, discovery] = await Promise.all([
    repositoryName(absRoot),
    discoverServices(absRoot),
  ]);

  const graph = new BlastGraph();
  for (const service of discovery.services) graph.addService(service);

  return {
    repository,
    root: absRoot,
    discovery: { strategy: discovery.strategy, attempted: discovery.attempted },
    graph,
    runtime: { connected: false },
  };
}

export function analysisToJSON(analysis: Analysis): AnalysisJSON {
  const { services, edges } = analysis.graph.toJSON();
  return {
    repository: analysis.repository,
    root: analysis.root,
    discovery: analysis.discovery,
    services,
    edges,
    runtime: analysis.runtime,
  };
}

/** `owner/repo` from the origin remote when there is one. */
export async function repositoryName(absRoot: string): Promise<string> {
  try {
    const { stdout } = await run("git", ["-C", absRoot, "remote", "get-url", "origin"], {
      timeout: 5_000,
    });
    const match = /[:/]([^/:\s]+)\/([^/\s]+?)(?:\.git)?\/?$/.exec(stdout.trim());
    if (match?.[1] && match[2]) return `${match[1]}/${match[2]}`;
  } catch {
    // not a git repository, or no origin: fall through
  }
  return path.basename(absRoot);
}

File: crates/blastradius-core/src/analyze.rs
//! Runs every built stage over a repository and returns the graph.
use std::path::{Path, PathBuf};
use std::process::Command;
use std::sync::LazyLock;

use regex::Regex;
use serde::{Deserialize, Serialize};
use thiserror::Error;

use crate::discover::{DiscoveryAttempt, discover_services};
use crate::fs::{FileIndex, is_dir};
use crate::graph::BlastGraph;
use crate::map::{self, MappingStats};
use crate::model::{DiscoveryStrategy, Edge, Service};

#[derive(Debug, Error)]
pub enum AnalyzeError {
    #[error("not a directory: {0}")]
    NotADirectory(String),
    #[error(transparent)]
    Io(#[from] std::io::Error),
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Discovery {
    pub strategy: Option<DiscoveryStrategy>,
    pub attempted: Vec<DiscoveryAttempt>,
}

/// Stage 3 fills this in. Until then the report says so plainly.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
pub struct Runtime {
    pub connected: bool,
}

#[derive(Debug, Clone)]
pub struct Analysis {
    /// `owner/repo` from the git remote, or the directory name.
    pub repository: String,
    /// Absolute path that was analysed.
    pub root: PathBuf,
    pub discovery: Discovery,
    pub graph: BlastGraph,
    pub mapping: MappingStats,
    pub runtime: Runtime,
}

/// The JSON contract. Field order is part of it.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AnalysisJson {
    pub repository: String,
    pub root: String,
    pub discovery: Discovery,
    pub services: Vec<Service>,
    pub edges: Vec<Edge>,
    pub mapping: MappingStats,
    pub runtime: Runtime,
}

impl Analysis {
    pub fn to_json(&self) -> AnalysisJson {
        AnalysisJson {
            repository: self.repository.clone(),
            root: self.root.display().to_string(),
            discovery: self.discovery.clone(),
            services: self.graph.services(),
            edges: self.graph.edges().to_vec(),
            mapping: self.mapping.clone(),
            runtime: self.runtime,
        }
    }
}

pub fn analyze(root: &Path) -> Result<Analysis, AnalyzeError> {
    if !is_dir(root) {
        return Err(AnalyzeError::NotADirectory(root.display().to_string()));
    }
    let root = root.canonicalize()?;
    let index = FileIndex::build(&root);
    let discovery = discover_services(&root, &index);

    let mapped = map::run(&root, &index, &discovery.services);

    let mut graph = BlastGraph::new();
    for service in discovery.services {
        graph.add_service(service);
    }
    for edge in mapped.edges {
        graph
            .add_edge(edge)
            .expect("mapping only produces edges between discovered services");
    }

    Ok(Analysis {
        repository: repository_name(&root),
        root,
        discovery: Discovery {
            strategy: discovery.strategy,
            attempted: discovery.attempted,
        },
        graph,
        mapping: mapped.stats,
        runtime: Runtime::default(),
    })
}

static REMOTE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"[:/]([^/:\s]+)/([^/\s]+?)(?:\.git)?/?$").unwrap());

/// `owner/repo` from the origin remote when there is one, else the
/// directory's name. The only subprocess the engine runs.
pub fn repository_name(root: &Path) -> String {
    let output = Command::new("git")
        .arg("-C")
        .arg(root)
        .args(["remote", "get-url", "origin"])
        .output();
    if let Ok(output) = output {
        if output.status.success() {
            let url = String::from_utf8_lossy(&output.stdout);
            if let Some(caps) = REMOTE.captures(url.trim()) {
                return format!("{}/{}", &caps[1], &caps[2]);
            }
        }
    }
    root.file_name()
        .map(|n| n.to_string_lossy().into_owned())
        .unwrap_or_default()
}
